crates/matrix-sdk-ui/src/sync_service/mod.rs

```diff
@@ -72,118 +104,303 @@ impl SyncService {
+                    None => {
+                        // The stream has ended.
+                        break (false, false);
                     }
                 }
-
-                state_observer.set(SyncServiceState::Terminated);
+            };
+
+            if let Err(err) = sender
+                .send(TerminationReport {
+                    is_error,
+                    has_expired,
+                    origin: TerminationOrigin::RoomList,
+                })
+                .await
+            {
+                error!("Error while sending termination report: {err:#}");
             }
-        }));
+        }
+    }
 
-        Ok(())
+    /// Start (or restart) the underlying sliding syncs.
+    ///
+    /// This can be called multiple times safely:
+    /// - if the stream is still properly running, it won't be restarted.
+    /// - if the stream has been aborted before, it will be properly cleaned up
+    ///   and restarted.
+    pub async fn start(&self) {
+        let _guard = self.modifying_state.lock().await;
+
+        // Only (re)start the tasks if any was stopped.
+        if matches!(self.state.get(), SyncServiceState::Running) {
+            // It was already true, so we can skip the restart.
+            return;
+        }
+
+        trace!("starting sync service");
+
+        let (sender, receiver) = tokio::sync::mpsc::channel(16);
+
+        // First, take care of the room list.
+        *self.room_list_task.lock().unwrap() =
+            Some(spawn(self.spawn_room_list_sync(sender.clone())));
+
+        // Then, take care of the encryption sync.
+        if let Some(encryption_sync) = self.encryption_sync.clone() {
+            *self.encryption_sync_task.lock().unwrap() =
+                Some(spawn(self.spawn_encryption_sync(encryption_sync, sender.clone())));
+        }
+
+        // Spawn the scheduler task.
+        *self.scheduler_sender.lock().unwrap() = Some(sender);
+        *self.scheduler_task.lock().unwrap() = Some(spawn(self.spawn_scheduler_task(receiver)));
+
+        self.state.set(SyncServiceState::Running);
     }
 
     /// Stop the underlying sliding syncs.
     ///
     /// This must be called when the app goes into the background. It's better
     /// to call this API when the application exits, although not strictly
     /// necessary.
-    pub fn pause(&self) -> Result<(), Error> {
-        self.room_list_service.stop_sync()?;
-        if let Some(ref encryption_sync) = self.encryption_sync {
-            encryption_sync.stop()?;
-        }
+    #[instrument(skip_all)]
+    pub async fn stop(&self) -> Result<(), Error> {
+        let _guard = self.modifying_state.lock().await;
+
+        match self.state.get() {
+            SyncServiceState::Idle | SyncServiceState::Terminated | SyncServiceState::Error => {
+                // No need to pause if we were not running.
+                return Ok(());
+            }
+            SyncServiceState::Running => {}
+        };
```

Comment:
```suggestion
        if matches!(self.state.get(), SyncServiceState::Running).not() {
            return Ok(());
        }
```

Reply:
I gently disagree here :grin: I prefer to see all the states explicitly and be warned by the compiler if at some point I add a new variant.